cpp/src/arrow/util/async_iterator.h

```diff
@@ -0,0 +1,295 @@
+    if (*finished) {
+      readahead_queue.push(Future<T>::MakeFinished(IterationTraits<T>::End()));
+    } else {
+      auto back_of_queue = source_generator();
+      back_of_queue.AddCallback(mark_finished_if_done);
+      readahead_queue.push(std::move(back_of_queue));
+    }
+    return result;
+  };
+}
+
+/// \brief Transforms an async generator using a transformer function returning a new
+/// AsyncGenerator
+///
+/// The transform function here behaves exactly the same as the transform function in
+/// MakeTransformedIterator and you can safely use the same transform function to
+/// transform both synchronous and asynchronous streams.
+template <typename T, typename V>
+AsyncGenerator<V> TransformAsyncGenerator(AsyncGenerator<T> generator,
+                                          Transformer<T, V> transformer) {
+  return TransformingGenerator<T, V>(generator, transformer);
+}
+
+namespace detail {
+
+template <typename T>
+struct BackgroundIteratorPromise : ReadaheadPromise {
+  ~BackgroundIteratorPromise() override {}
+
+  explicit BackgroundIteratorPromise(Iterator<T>* it) : it_(it) {}
+
+  bool Call() override {
+    auto next = it_->Next();
+    auto finished = next == IterationTraits<T>::End();
+    out_.MarkFinished(std::move(next));
+    return finished;
+  }
+
+  void End() override { out_.MarkFinished(IterationTraits<T>::End()); }
+
+  Iterator<T>* it_;
+  Future<T> out_ = Future<T>::Make();
+};
+
+}  // namespace detail
+
+/// \brief Async generator that iterates on an underlying iterator in a
+/// separate thread.
+template <typename T>
+class BackgroundIterator {
+  using PromiseType = typename detail::BackgroundIteratorPromise<T>;
+
+ public:
+  explicit BackgroundIterator(Iterator<T> it, internal::Executor* executor)
+      : it_(new Iterator<T>(std::move(it))),
+        queue_(new detail::ReadaheadQueue(0)),
+        executor_(executor),
+        done_() {}
+
+  ~BackgroundIterator() {
+    if (queue_) {
+      // Make sure the queue doesn't call any promises after this object
+      // is destroyed.
+      queue_->EnsureShutdownOrDie();
+    }
+  }
+
+  ARROW_DEFAULT_MOVE_AND_ASSIGN(BackgroundIterator);
+  ARROW_DISALLOW_COPY_AND_ASSIGN(BackgroundIterator);
+
+  Future<T> operator()() {
+    if (done_) {
+      return Future<T>::MakeFinished(IterationTraits<T>::End());
+    }
+    auto promise = std::unique_ptr<PromiseType>(new PromiseType{it_.get()});
+    auto result = Future<T>(promise->out_);
+    // TODO: Need a futuristic version of ARROW_RETURN_NOT_OK
+    auto append_status = queue_->Append(
+        static_cast<std::unique_ptr<detail::ReadaheadPromise>>(std::move(promise)));
+    if (!append_status.ok()) {
+      return Future<T>::MakeFinished(append_status);
+    }
+
+    result.AddCallback([this](const Result<T>& result) {
+      if (!result.ok() || result.ValueUnsafe() == IterationTraits<T>::End()) {
+        done_ = true;
+      }
+    });
+
+    return executor_->Transfer(result);
```

Comment:
Actually, that trivial callback added a few lines above will not run on the executor.  It will run on the I/O background thread.  Instead any continuation added (by the caller) to the returned future will be run on the executor.  This is how we transfer control from the background I/O thread to the CPU executor.

Reply:
But is it the background generator's job to call `Transfer`? Is the caller simply meant to pass the "executor of the current thread"?

It seems at some point we should add the notion of a per-thread currently running executor. What do you think?

Reply:
I think that is a good idea.  Then we could have a simple mandate: all functions that return a future should return a future that continues on the caller's executor.

Reply:
Hmm... I'm not sure. Can we discuss this with Ben (e.g. on Zulip)? It would seem more logical to me if `Then` would by default schedule the callback on the current executor (that is, the executor of the thread calling `Then`).